File: app/src/main/java/com/krystianwsul/checkme/firebase/RemoteProjectFactory.java
package com.krystianwsul.checkme.firebase;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.google.firebase.database.DataSnapshot;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.domainmodel.UserInfo;
import com.krystianwsul.checkme.domainmodel.local.LocalCustomTime;
import com.krystianwsul.checkme.firebase.json.CustomTimeJson;
import com.krystianwsul.checkme.firebase.json.JsonWrapper;
import com.krystianwsul.checkme.firebase.json.ProjectJson;
import com.krystianwsul.checkme.firebase.json.TaskJson;
import com.krystianwsul.checkme.firebase.json.UserJson;
import com.krystianwsul.checkme.firebase.records.RemoteProjectManager;
import com.krystianwsul.checkme.firebase.records.RemoteProjectRecord;
import com.krystianwsul.checkme.loaders.CreateTaskLoader;
import com.krystianwsul.checkme.utils.CustomTimeKey;
import com.krystianwsul.checkme.utils.InstanceKey;
import com.krystianwsul.checkme.utils.TaskKey;
import com.krystianwsul.checkme.utils.time.DayOfWeek;
import com.krystianwsul.checkme.utils.time.ExactTimeStamp;

import junit.framework.Assert;

import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class RemoteProjectFactory {
    @NonNull
    private final DomainFactory mDomainFactory;

    @NonNull
    private final UserInfo mUserInfo;

    @NonNull
    private final RemoteProjectManager mRemoteProjectManager;

    @NonNull
    private final Map<String, RemoteProject> mRemoteProjects;

    @NonNull
    private final String mUuid;

    public RemoteProjectFactory(@NonNull DomainFactory domainFactory, @NonNull Iterable<DataSnapshot> children, @NonNull UserInfo userInfo, @NonNull String uuid, @NonNull ExactTimeStamp now) {
        mDomainFactory = domainFactory;
        mUserInfo = userInfo;
        mUuid = uuid;

        mRemoteProjectManager = new RemoteProjectManager(domainFactory, children);

        mRemoteProjects = Stream.of(mRemoteProjectManager.getRemoteProjectRecords().values())
                .map(remoteProjectRecord -> new RemoteProject(domainFactory, remoteProjectRecord, mUserInfo, uuid, now))
                .collect(Collectors.toMap(RemoteProject::getId, remoteProject -> remoteProject));
    }

    @NonNull
    public RemoteTask createScheduleRootTask(@NonNull ExactTimeStamp now, @NonNull String name, @NonNull List<CreateTaskLoader.ScheduleData> scheduleDatas, @Nullable String note, @NonNull String projectId) {
        RemoteTask remoteTask = createRemoteTaskHelper(now, name, note, projectId);

        remoteTask.createSchedules(now, scheduleDatas);

        return remoteTask;
    }

    @NonNull
    public RemoteTask createRemoteTaskHelper(@NonNull ExactTimeStamp now, @NonNull String name, @Nullable String note, @NonNull String projectId) {
        TaskJson taskJson = new TaskJson(name, now.getLong(), null, null, null, null, note, Collections.emptyMap());

        return getRemoteProjectForce(projectId).newRemoteTask(taskJson, now);
    }

    @NonNull
    public RemoteProject createRemoteProject(@NonNull String name, @NonNull ExactTimeStamp now, @NonNull Set<String> recordOf, @NonNull RemoteRootUser remoteRootUser) {
        Assert.assertTrue(!TextUtils.isEmpty(name));

        Set<String> friendIds = new HashSet<>(recordOf);
        friendIds.remove(mUserInfo.getKey());

        Map<String, UserJson> userJsons = mDomainFactory.getUserJsons(friendIds);
        userJsons.put(mUserInfo.getKey(), remoteRootUser.getUserJson());

        ProjectJson projectJson = new ProjectJson(name, now.getLong(), null, new HashMap<>(), new HashMap<>(), new HashMap<>(), userJsons);

        RemoteProjectRecord remoteProjectRecord = mRemoteProjectManager.newRemoteProjectRecord(mDomainFactory, new JsonWrapper(recordOf, projectJson));

        RemoteProject remoteProject = new RemoteProject(mDomainFactory, remoteProjectRecord, mUserInfo, mUuid, now);

        Assert.assertTrue(!mRemoteProjects.containsKey(remoteProject.getId()));

        mRemoteProjects.put(remoteProject.getId(), remoteProject);

        return remoteProject;
    }

    public void save() {
        Assert.assertTrue(!mRemoteProjectManager.isSaved());

        mRemoteProjectManager.save();
    }

    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    public boolean isSaved() {
        return mRemoteProjectManager.isSaved();
    }

    @NonNull
    public Stream<RemoteTask> getTasks() {
        return Stream.of(mRemoteProjects.values())
                .map(RemoteProject::getRemoteTasks)
                .flatMap(Stream::of);
    }

    @NonNull
    String getRemoteCustomTimeId(@NonNull CustomTimeKey customTimeKey, @NonNull RemoteProject remoteProject) {
        Assert.assertTrue(customTimeKey.mLocalCustomTimeId != null);
        Assert.assertTrue(TextUtils.isEmpty(customTimeKey.mRemoteCustomTimeId));

        int localCustomTimeId = customTimeKey.mLocalCustomTimeId;

        LocalCustomTime localCustomTime = mDomainFactory.getLocalFactory().getLocalCustomTime(localCustomTimeId);

        if (!localCustomTime.hasRemoteRecord(remoteProject.getId())) {
            CustomTimeJson customTimeJson = new CustomTimeJson(mDomainFactory.getLocalFactory().getUuid(), localCustomTime.getId(), localCustomTime.getName(), localCustomTime.getHourMinute(DayOfWeek.SUNDAY).getHour(), localCustomTime.getHourMinute(DayOfWeek.SUNDAY).getMinute(), localCustomTime.getHourMinute(DayOfWeek.MONDAY).getHour(), localCustomTime.getHourMinute(DayOfWeek.MONDAY).getMinute(), localCustomTime.getHourMinute(DayOfWeek.TUESDAY).getHour(), localCustomTime.getHourMinute(DayOfWeek.TUESDAY).getMinute(), localCustomTime.getHourMinute(DayOfWeek.WEDNESDAY).getHour(), localCustomTime.getHourMinute(DayOfWeek.WEDNESDAY).getMinute(), localCustomTime.getHourMinute(DayOfWeek.THURSDAY).getHour(), localCustomTime.getHourMinute(DayOfWeek.THURSDAY).getMinute(), localCustomTime.getHourMinute(DayOfWeek.FRIDAY).getHour(), localCustomTime.getHourMinute(DayOfWeek.FRIDAY).getMinute(), localCustomTime.getHourMinute(DayOfWeek.SATURDAY).getHour(), localCustomTime.getHourMinute(DayOfWeek.SATURDAY).getMinute());

            RemoteCustomTime remoteCustomTime = remoteProject.newRemoteCustomTime(customTimeJson);

            localCustomTime.addRemoteCustomTimeRecord(remoteCustomTime.getRemoteCustomTimeRecord());
        }

        return localCustomTime.getRemoteId(remoteProject.getId());
    }

    @NonNull
    public RemoteCustomTime getRemoteCustomTime(@NonNull String remoteProjectId, @NonNull String remoteCustomTimeId) {
        Assert.assertTrue(!TextUtils.isEmpty(remoteProjectId));
        Assert.assertTrue(!TextUtils.isEmpty(remoteCustomTimeId));

        Assert.assertTrue(mRemoteProjects.containsKey(remoteProjectId));

        RemoteProject remoteProject = mRemoteProjects.get(remoteProjectId);
        Assert.assertTrue(remoteProject != null);

        return remoteProject.getRemoteCustomTime(remoteCustomTimeId);
    }

    @NonNull
    public List<RemoteCustomTime> getRemoteCustomTimes() {
        return Stream.of(mRemoteProjects.values())
                .map(RemoteProject::getRemoteCustomTimes)
                .map(Stream::of)
                .flatMap(stream -> stream)
                .collect(Collectors.toList());
    }

    public int getInstanceCount() {
        return Stream.of(mRemoteProjects.values())
                .map(RemoteProject::getRemoteTasks)
                .flatMap(Stream::of)
                .map(remoteTask -> remoteTask.getExistingInstances().size())
                .reduce(0, (x, y) -> x + y);
    }

    @NonNull
    public List<RemoteInstance> getExistingInstances() {
        return Stream.of(mRemoteProjects.values())
                .map(RemoteProject::getRemoteTasks).flatMap(Stream::of).flatMap(remoteTask -> Stream.of(remoteTask.getExistingInstances().values())).map(instance -> (RemoteInstance) instance)
                .collect(Collectors.toList());
    }

    @Nullable
    public RemoteInstance getExistingInstanceIfPresent(@NonNull InstanceKey instanceKey) {
        TaskKey taskKey = instanceKey.mTaskKey;

        if (TextUtils.isEmpty(taskKey.mRemoteTaskId))
            return null;

        RemoteTask remoteTask = getRemoteProjectForce(taskKey).getRemoteTaskIfPresent(taskKey.mRemoteTaskId);
        if (remoteTask == null)
            return null;

        return remoteTask.getExistingInstanceIfPresent(instanceKey.mScheduleKey);
    }

    @NonNull
    private RemoteProject getRemoteProjectForce(@NonNull TaskKey taskKey) {
        RemoteProject remoteProject = getRemoteProjectIfPresent(taskKey);
        Assert.assertTrue(remoteProject != null);

        return remoteProject;
    }

    @Nullable
    private RemoteProject getRemoteProjectIfPresent(@NonNull TaskKey taskKey) {
        Assert.assertTrue(!TextUtils.isEmpty(taskKey.mRemoteProjectId));
        Assert.assertTrue(!TextUtils.isEmpty(taskKey.mRemoteTaskId));

        return mRemoteProjects.get(taskKey.mRemoteProjectId);
    }

    @NonNull
    public RemoteTask getTaskForce(@NonNull TaskKey taskKey) {
        Assert.assertTrue(!TextUtils.isEmpty(taskKey.mRemoteTaskId));

        return getRemoteProjectForce(taskKey).getRemoteTaskForce(taskKey.mRemoteTaskId);
    }

    @Nullable
    public RemoteTask getTaskIfPresent(@NonNull TaskKey taskKey) {
        Assert.assertTrue(!TextUtils.isEmpty(taskKey.mRemoteTaskId));

        RemoteProject remoteProject = getRemoteProjectIfPresent(taskKey);
        if (remoteProject == null)
            return null;

        return remoteProject.getRemoteTaskIfPresent(taskKey.mRemoteTaskId);
    }

    @NonNull
    public Set<TaskKey> getTaskKeys() {
        Set<TaskKey> taskKeys = new HashSet<>();

        for (RemoteProject remoteProject : mRemoteProjects.values()) {
            Assert.assertTrue(remoteProject != null);

            String projectId = remoteProject.getId();

            taskKeys.addAll(Stream.of(remoteProject.getTaskIds())
                    .map(taskId -> new TaskKey(projectId, taskId))
                    .collect(Collectors.toList()));
        }

        return taskKeys;
    }

    public int getTaskCount() {
        return Stream.of(mRemoteProjects.values())
                .map(RemoteProject::getRemoteTasks)
                .map(Collection::size)
                .reduce(0, (x, y) -> x + y);
    }

    @NonNull
    Set<RemoteTaskHierarchy> getTaskHierarchiesByChildTaskKey(@NonNull TaskKey childTaskKey) {
        Assert.assertTrue(!TextUtils.isEmpty(childTaskKey.mRemoteTaskId));

        return getRemoteProjectForce(childTaskKey).getTaskHierarchiesByChildTaskKey(childTaskKey);
    }

    @NonNull
    public Collection<RemoteProject> getRemoteProjects() {
        return mRemoteProjects.values();
    }

    public void updateUserInfo(@NonNull UserInfo userInfo) {
        Stream.of(mRemoteProjects.values())
                .forEach(remoteProject -> remoteProject.updateUserInfo(userInfo, mUuid));
    }

    @NonNull
    public RemoteProject getRemoteProjectForce(@NonNull String projectId) {
        Assert.assertTrue(!TextUtils.isEmpty(projectId));
        Assert.assertTrue(mRemoteProjects.containsKey(projectId));

        RemoteProject remoteProject = mRemoteProjects.get(projectId);
        Assert.assertTrue(remoteProject != null);

        return remoteProject;
    }

    void deleteProject(@NonNull RemoteProject remoteProject) {
        String projectId = remoteProject.getId();

        Assert.assertTrue(mRemoteProjects.containsKey(projectId));
        mRemoteProjects.remove(projectId);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/RemoteProjectFactory.kt
package com.krystianwsul.checkme.firebase

import android.text.TextUtils
import com.google.firebase.database.DataSnapshot
import com.krystianwsul.checkme.domainmodel.DomainFactory
import com.krystianwsul.checkme.domainmodel.local.LocalFactory
import com.krystianwsul.checkme.firebase.managers.AndroidRemotePrivateProjectManager
import com.krystianwsul.checkme.firebase.managers.AndroidRemoteSharedProjectManager
import com.krystianwsul.common.domain.DeviceDbInfo
import com.krystianwsul.common.domain.DeviceInfo
import com.krystianwsul.common.domain.UserInfo
import com.krystianwsul.common.firebase.json.JsonWrapper
import com.krystianwsul.common.firebase.json.SharedProjectJson
import com.krystianwsul.common.firebase.json.TaskJson
import com.krystianwsul.common.firebase.models.*
import com.krystianwsul.common.firebase.records.RemoteTaskRecord
import com.krystianwsul.common.time.ExactTimeStamp
import com.krystianwsul.common.time.Time
import com.krystianwsul.common.utils.*

class RemoteProjectFactory(
        deviceDbInfo: DeviceDbInfo,
        private val localFactory: LocalFactory,
        sharedChildren: List<DataSnapshot>,
        privateSnapshot: DataSnapshot,
        now: ExactTimeStamp
) : Project.Parent {

    private val remotePrivateProjectManager = AndroidRemotePrivateProjectManager(deviceDbInfo.userInfo, privateSnapshot, now)
    private val remoteSharedProjectManager = AndroidRemoteSharedProjectManager(sharedChildren)

    var remotePrivateProject = PrivateProject(remotePrivateProjectManager.remoteProjectRecord).apply { fixNotificationShown(localFactory, now) }
        private set

    val remoteSharedProjects = remoteSharedProjectManager.remoteProjectRecords
            .values
            .associate { (remoteSharedProjectRecord, _) ->
                remoteSharedProjectRecord.id to SharedProject(remoteSharedProjectRecord).apply {
                    fixNotificationShown(localFactory, now)
                    updateDeviceDbInfo(deviceDbInfo)
                }
            }
            .toMutableMap()

    val remoteProjects
        get() = remoteSharedProjects.toMutableMap<ProjectKey, Project<*, *>>().apply {
            put(remotePrivateProject.id, remotePrivateProject)
        }.toMap()

    var isPrivateSaved
        get() = remotePrivateProjectManager.isSaved
        set(value) {
            remotePrivateProjectManager.isSaved = value
        }

    val isSharedSaved get() = remoteSharedProjectManager.isSaved

    val eitherSaved get() = isPrivateSaved || isSharedSaved

    val tasks get() = remoteProjects.values.flatMap { it.tasks }

    val remoteCustomTimes get() = remoteProjects.values.flatMap { it.customTimes }

    val instanceCount
        get() = remoteProjects.values
                .flatMap { it.tasks }
                .asSequence()
                .map { it.existingInstances.size }
                .sum()

    val existingInstances
        get() = remoteProjects.values.flatMap { it.existingInstances }

    val taskKeys
        get() = remoteProjects.values
                .flatMap {
                    val projectId = it.id

                    it.taskIds.map { TaskKey(projectId, it) }
                }
                .toSet()

    val taskCount: Int
        get() = remoteProjects.values
                .map { it.tasks.size }
                .sum()

    fun onChildEvent(deviceDbInfo: DeviceDbInfo, databaseEvent: DatabaseEvent, now: ExactTimeStamp): Boolean {
        val projectKey = ProjectKey.Shared(databaseEvent.key)
        val projectPair = remoteSharedProjectManager.remoteProjectRecords[projectKey]

        if (projectPair?.second == true) {
            remoteSharedProjectManager.remoteProjectRecords[projectKey] = Pair(projectPair.first, false)

            return true
        } else {
            when (databaseEvent) {
                is DatabaseEvent.AddChange -> {
                    try {
                        val remoteProjectRecord = remoteSharedProjectManager.setChild(databaseEvent.dataSnapshot)

                        remoteSharedProjects[remoteProjectRecord.id] = SharedProject(remoteProjectRecord).apply {
                            fixNotificationShown(localFactory, now)
                            updateDeviceDbInfo(deviceDbInfo)
                        }
                    } catch (onlyVisibilityPresentException: RemoteTaskRecord.OnlyVisibilityPresentException) {
                        // hack for oldestVisible being set on records removed by cloud function
                    }
                }
                is DatabaseEvent.Remove -> {
                    remoteSharedProjectManager.removeChild(projectKey)
                    remoteSharedProjects.remove(projectKey)
                }
            }

            return false
        }
    }

    fun onNewPrivate(dataSnapshot: DataSnapshot, now: ExactTimeStamp) {
        try {
            val remotePrivateProjectRecord = remotePrivateProjectManager.newSnapshot(dataSnapshot)

            remotePrivateProject = PrivateProject(remotePrivateProjectRecord).apply { fixNotificationShown(localFactory, now) }
        } catch (onlyVisibilityPresentException: RemoteTaskRecord.OnlyVisibilityPresentException) {
            // hack for oldestVisible being set on records removed by cloud function

            /*
            todo: either set oldestVisible once you're sure the database has been updated, or find
              a better mechanism than oldestVisible
             */
        }
    }

    fun createScheduleRootTask(
            now: ExactTimeStamp,
            name: String,
            scheduleDatas: List<Pair<ScheduleData, Time>>,
            note: String?,
            projectId: ProjectKey,
            imageUuid: String?,
            deviceDbInfo: DeviceDbInfo
    ) = createRemoteTaskHelper(now, name, note, projectId, imageUuid, deviceDbInfo).apply {
        createSchedules(remotePrivateProject.id.toUserKey(), now, scheduleDatas)
    }

    fun createRemoteTaskHelper(
            now: ExactTimeStamp,
            name: String,
            note: String?,
            projectId: ProjectKey,
            imageUuid: String?,
            deviceDbInfo: DeviceDbInfo
    ): RemoteTask<*, *> {
        val image = imageUuid?.let { TaskJson.Image(imageUuid, deviceDbInfo.uuid) }
        val taskJson = TaskJson(name, now.long, null, note, image = image)

        return getRemoteProjectForce(projectId).newRemoteTask(taskJson)
    }

    fun createRemoteProject(
            name: String,
            now: ExactTimeStamp,
            recordOf: Set<UserKey>,
            remoteRootUser: RemoteRootUser,
            userInfo: UserInfo,
            remoteFriendFactory: RemoteFriendFactory
    ): SharedProject {
        check(!TextUtils.isEmpty(name))

        val friendIds = recordOf.toMutableSet()
        friendIds.remove(userInfo.key)

        val userJsons = remoteFriendFactory.getUserJsons(friendIds)
        userJsons[userInfo.key] = remoteRootUser.userJson

        val projectJson = SharedProjectJson(
                name,
                now.long,
                users = userJsons.mapKeys { it.key.key }.toMutableMap()
        )

        val remoteProjectRecord = remoteSharedProjectManager.newRemoteProjectRecord(JsonWrapper(projectJson))

        val remoteProject = SharedProject(remoteProjectRecord)

        check(!remoteProjects.containsKey(remoteProject.id))

        remoteSharedProjects[remoteProject.id] = remoteProject

        return remoteProject
    }

    fun save(domainFactory: DomainFactory): Boolean {
        val privateSaved = remotePrivateProjectManager.save(domainFactory)
        val sharedSaved = remoteSharedProjectManager.save(domainFactory)
        return privateSaved || sharedSaved
    }

    fun getRemoteCustomTime(remoteProjectId: ProjectKey, remoteCustomTimeId: RemoteCustomTimeId): CustomTime<*, *> {
        check(remoteProjects.containsKey(remoteProjectId))

        return remoteProjects.getValue(remoteProjectId).getRemoteCustomTime(remoteCustomTimeId)
    }

    fun getExistingInstanceIfPresent(instanceKey: InstanceKey): RemoteInstance<*, *>? {
        val taskKey = instanceKey.taskKey

        if (TextUtils.isEmpty(taskKey.remoteTaskId))
            return null

        val remoteTask = getRemoteProjectForce(taskKey).getRemoteTaskIfPresent(taskKey.remoteTaskId)
                ?: return null

        return remoteTask.getExistingInstanceIfPresent(instanceKey.scheduleKey)
    }

    private fun getRemoteProjectForce(taskKey: TaskKey) = getRemoteProjectIfPresent(taskKey)!!

    private fun getRemoteProjectIfPresent(taskKey: TaskKey): Project<*, *>? {
        check(!TextUtils.isEmpty(taskKey.remoteTaskId))

        return remoteProjects[taskKey.remoteProjectId]
    }

    fun getTaskForce(taskKey: TaskKey): RemoteTask<*, *> {
        check(!TextUtils.isEmpty(taskKey.remoteTaskId))

        return getRemoteProjectForce(taskKey).getRemoteTaskForce(taskKey.remoteTaskId)
    }

    fun getTaskIfPresent(taskKey: TaskKey): RemoteTask<*, *>? {
        check(!TextUtils.isEmpty(taskKey.remoteTaskId))

        return getRemoteProjectIfPresent(taskKey)?.getRemoteTaskIfPresent(taskKey.remoteTaskId)
    }

    fun updateDeviceInfo(deviceDbInfo: DeviceDbInfo) = remoteSharedProjects.values.forEach { it.updateDeviceDbInfo(deviceDbInfo) }

    fun updatePhotoUrl(deviceInfo: DeviceInfo, photoUrl: String) = remoteSharedProjects.values.forEach { it.updatePhotoUrl(deviceInfo, photoUrl) }

    fun getRemoteProjectForce(projectId: ProjectKey): Project<*, *> {
        check(remoteProjects.containsKey(projectId))

        return remoteProjects.getValue(projectId)
    }

    fun getRemoteProjectIfPresent(projectId: String) = remoteProjects.entries
            .singleOrNull { it.key.key == projectId }
            ?.value

    override fun deleteProject(project: Project<*, *>) {
        val projectId = project.id

        check(remoteProjects.containsKey(projectId))
        remoteSharedProjects.remove(projectId)
    }

    fun getTaskHierarchy(remote: TaskHierarchyKey.Remote) = remoteProjects.getValue(remote.projectId).getTaskHierarchy(remote.taskHierarchyId)

    fun getSchedule(scheduleId: ScheduleId.Remote) = remoteProjects.getValue(scheduleId.projectId).getRemoteTaskForce(scheduleId.taskId)
            .schedules
            .single { it.scheduleId == scheduleId }

}
